### ISSUE TITLE
[HOLD for Payment 2024-07-22] [$250] Improve ReportScreen rendering time

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#44925]
### Problem
The Profiler indicates that the ReportScreen render time gets re-rendered dozens of times, with each re-rendering taking an average of 599.8ms.

The most current problems are with [this hook](https://github.com/Expensify/App/blob/c80e8b65eb8bb5f218d50a59efb2a65f22d7c527/src/pages/home/ReportScreen.tsx#L173):
```
const lastAccessedReportID = useLastAccessedReportID(!!route.params.openOnAdminRoom);
```

In [this PR](https://github.com/Expensify/App/pull/44559) we’ve introduced the above hook and there might be several problems with it:

1. We are calculating `lastAccessedReportID` in each re-render, however, we only require it in one particular scenario: when the `route.params.reportID` is empty,
2. The `useLastAccessedReportID` hook is heavy, due to the poor performance of the `ReportUtils.findLastAccessedReport` function. This issue existed before the introduction of the PR mentioned.

### Solution

Ad 1.
We should calculate the lastAccessedReportID only when needed, that means only once, and only when route.params.reportID is empty ([code](https://github.com/Expensify/App/blob/ce5ae736eccd5180f50e47e21db047dd33b90a43/src/pages/home/ReportScreen.tsx#L176)). This value is used to determine which chat we should show to the user when the user accesses the website without reportID in the URL. [POC](https://github.com/Expensify/App/commit/5352992d30f919b6b66ed2add60b1ec6b41ca143)

Ad 2.
After diving deep into the code I’ve discovered that [we sort a whole array of reports](https://github.com/Expensify/App/blob/c80e8b65eb8bb5f218d50a59efb2a65f22d7c527/src/libs/ReportUtils.ts#L1194) just to get [one, the most recent, report](https://github.com/Expensify/App/blob/c80e8b65eb8bb5f218d50a59efb2a65f22d7c527/src/libs/ReportUtils.ts#L1245). We can replace this sort with `lodashMaxBy`, as we do not (from my analysis) need to sort all reports (needs confirmation). [POC](https://github.com/Expensify/App/commit/41dff749202c7c11eb6d74ed57f69a78cc61ca2b)

cc @kosmydel 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0109e0b5642eb1863c</li>
        <li>Upwork Job ID: 1809262837407973619</li>
        <li>Last Price Increase: 2024-07-05</li>
<li>Automatic offers: </li>
<ul>
<li>ikevin127 | Reviewer | 103028165</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @zanyrenney</details>

**Comment by kosmydel** [comment]
Hey, I’m from Software Mansion and I will take care of this issue :)

**Comment by mountiny** [comment]
A discussed in Slack, please try to split this up to two PRs

**Comment by ikevin127** [comment]
⚠️ Automation failed here -> this should be on **[HOLD for Payment 2024-07-22]** according to today's **production** deploy from https://github.com/Expensify/App/pull/44948#issuecomment-2229340753 and https://github.com/Expensify/App/pull/44955#issuecomment-2229341167.

I'm the only one requiring payment here for reviewing PRs https://github.com/Expensify/App/pull/44948 and https://github.com/Expensify/App/pull/44955. Additionally, 1 regression was reported above, therefore bounty payment shoud be adjusted to match that - waiting for regression confirmation from @mountiny.

cc @zanyrenney

**Comment by mountiny** [comment]
$375 to @ikevin127 for reviewing 2 PRs and deducing 50% of one bounty for the regression

**Comment by ikevin127** [comment]
@zanyrenney Friendly bump for payments, as I noticed this doensn't have the `Daily` label which could cause delays.

**Comment by zanyrenney** [comment]
**payment summary**

@ikevin127 - paid $375 via upwork
@kosmydel - internal contractor - not paid incrementally. 

Thanks!
